lib/Conversion/ConvertToArcs/ConvertToArcs.cpp

```diff
@@ -0,0 +1,392 @@
+          }
+        }
+        if (anyExternal) {
+          // LLVM_DEBUG(llvm::dbgs() << "    - Adding " <<
+          // result.getType()
+          //                         << " output for result "
+          //                         << result.getResultNumber() <<
+          //                         "\n");
+          outputs.push_back(result);
+          outputTypes.push_back(result.getType());
+        }
+      }
+    }
+    assert(lastOp);
+    builder.create<arc::OutputOp>(lastOp->getLoc(), outputs);
+
+    // Create the arc definition.
+    builder.setInsertionPoint(module);
+    auto defOp = builder.create<DefineOp>(
+        lastOp->getLoc(),
+        builder.getStringAttr(
+            globalNamespace.newName(module.moduleName() + "_arc")),
+        builder.getFunctionType(inputTypes, outputTypes));
+    defOp.getBody().push_back(block.release());
+
+    // Create the call to the arc definition to replace the operations that
+    // we have just extracted.
+    builder.setInsertionPoint(module.getBodyBlock()->getTerminator());
+    auto arcOp = builder.create<StateOp>(lastOp->getLoc(), defOp, Value{},
+                                         Value{}, 0, inputs);
+    arcUses.push_back(arcOp);
+    for (auto [use, resultIdx] : externalUses)
+      use->set(arcOp.getResult(resultIdx));
+  }
+}
+
+void Converter::absorbRegs(HWModuleOp module) {
+  // Handle the trivial cases where all of an arc's results are used by
+  // exactly one register each.
+  unsigned outIdx = 0;
+  unsigned numTrivialRegs = 0;
+  for (auto &arc : arcUses) {
+    Value clock = arc.getClock();
+    SmallVector<seq::CompRegOp> absorbedRegs;
+    SmallVector<Attribute> absorbedNames(arc.getNumResults(), {});
+    if (auto names = arc->getAttrOfType<ArrayAttr>("names"))
+      absorbedNames.assign(names.getValue().begin(), names.getValue().end());
+    bool isTrivial = true;
+    for (auto result : arc.getResults()) {
+      if (!result.hasOneUse()) {
+        isTrivial = false;
+        break;
+      }
+      auto &use = *result.use_begin();
+      auto regOp = dyn_cast<seq::CompRegOp>(use.getOwner());
+      if (!regOp || regOp.getInput() != result ||
+          (clock && clock != regOp.getClk())) {
+        isTrivial = false;
+        break;
+      }
+      clock = regOp.getClk();
+      absorbedRegs.push_back(regOp);
+      absorbedNames[result.getResultNumber()] = regOp.getNameAttr();
+    }
+    if (!isTrivial) {
+      arcUses[outIdx++] = arc;
+      continue;
+    }
+    ++numTrivialRegs;
+
+    OpBuilder builder(arc);
+    auto newArc = builder.create<StateOp>(
+        arc.getLoc(), arc.getArcAttr(), arc.getResultTypes(), clock, Value{},
+        arc.getLatency() + 1, arc.getOperands());
+    newArc->setAttr("names", builder.getArrayAttr(absorbedNames));
+    for (auto [arcResult, reg] : llvm::zip(newArc.getResults(), absorbedRegs)) {
+      reg.replaceAllUsesWith(arcResult);
+      reg.erase();
+      auto it = arcBreakerIndices.find(reg);
```

Comment:
I guess this is causing UAF.
```suggestion
      auto it = arcBreakerIndices.find(reg);
      reg.erase();
```